<h1> <i> <b>
Name: AASHI AGGARWAL <br>
ID: 8920299

</h1> </i> </b>

In [99]:
import numpy as np
from sklearn.model_selection import train_test_split 
from sklearn.metrics import mean_squared_error 
from sklearn.linear_model import LinearRegression
from sklearn import metrics
import matplotlib.pyplot as plt  
from sklearn import datasets
import pandas as pd

<h3>
1. LOAD DATA--> Diabetes dataset <br>
2. Extracting X which is independent variable <br>
    - to extract all the features to input the model <br>
3. Extracting Y which is the dependent variable <br>
    - to compare the predicted and actual model
4. Splitting data into training and testing sets <br/>
- 
</h3>

In [100]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split

# Load the Diabetes dataset
diabetes = load_diabetes(scaled=True)

# Extract features (x) and target (y)
x,y = diabetes.data , diabetes.target



x_train, x_temp, y_train, y_temp = train_test_split(x, y, test_size=0.4, random_state=42)
x_val, x_test, y_val, y_test = train_test_split(x_temp, y_temp, test_size=0.5, random_state=42)

<h3>
5. Run a multivariate linear regression on all variables
</h3>

In [101]:
from sklearn.linear_model import LinearRegression

# Initialize and train the linear regression model
lin_reg = LinearRegression()
lin_reg.fit(x_train, y_train)


LinearRegression()

<h3> 
6.  Run a polynomial regression of the 2nd degree on the BMI feature alone <br>

<br>

</h3>

In [102]:
from sklearn.preprocessing import PolynomialFeatures

df= pd.DataFrame(data=diabetes.data,columns=diabetes.feature_names)

#BMI feature
x_train_bmi = np.array(df["bmi"]).reshape((-1, 1))
# PolynomialFeatures
poly = PolynomialFeatures(degree=2)
x_train_bmi_poly = poly.fit_transform(x_train_bmi)




<h3>
7. Run a multivariate polynomial regression of the 2nd degree on all variable <br>
#setting bias - false <br>

</h3>


In [103]:
poly = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly.fit_transform(x_train)

<h3>
Compare the three models by looking at MAE and MAPE using the provided function

<b> ----------------</b>

<br>
<h2> 
1. Predicting each model



</h2>

In [104]:
from sklearn.metrics import r2_score, mean_absolute_error
import numpy as np

y_pred_lin_reg = lin_reg.predict(x_test)
y_pred_bmi_poly = np.dot(x_val[:, 2:3] ** 2, lin_reg.coef_[2:3]) + np.dot(x_val[:, 2:3], lin_reg.coef_[2:3]) + lin_reg.intercept_
y_pred_multi_poly = lin_reg.predict(x_val)



<h2> 
2. Calculating R-squared for each model. <br>




</h2>

In [105]:
# Calculate R-squared for each model
r2_lin_reg = r2_score(y_test, y_pred_lin_reg)
r2_bmi_poly = r2_score(y_val, y_pred_bmi_poly)
r2_multi_poly = r2_score(y_val, y_pred_multi_poly)

print("R-squared (Linear Regression):", r2_lin_reg)
print("R-squared (BMI Polynomial Regression):", r2_bmi_poly)
print("R-squared (Multivariate Polynomial Regression):", r2_multi_poly)



R-squared (Linear Regression): 0.43891375898156637
R-squared (BMI Polynomial Regression): 0.31378440391273243
R-squared (Multivariate Polynomial Regression): 0.5809667034070143


<h1> Based on R - squared value:
<br>
As, higher R-squared indicates that the model explains a larger portion of the variance in the target variable.
Multivariate Polynomial Regression models perform better and explain a larger proportion of the variance in the target variable which makes them better as compared to other Regression model.</h1>

<br>


<h2> 
2. Calculating MAPE for each model. <br>
#Mean Absolute Percentage Error
</h2>


In [106]:

# Calculate MAPE for each model
def calculate_mape(y_true, y_pred):
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

mape_lin_reg = calculate_mape(y_test, y_pred_lin_reg)
mape_bmi_poly = calculate_mape(y_val, y_pred_bmi_poly)
mape_multi_poly = calculate_mape(y_val, y_pred_multi_poly)


print("\nMAPE (Linear Regression):", mape_lin_reg, "%")
print("MAPE (BMI Polynomial Regression):", mape_bmi_poly, "%")
print("MAPE (Multivariate Polynomial Regression):", mape_multi_poly, "%")


MAPE (Linear Regression): 34.92031805723895 %
MAPE (BMI Polynomial Regression): 50.3700071945691 %
MAPE (Multivariate Polynomial Regression): 34.80184412774623 %


<h1> Based on MAPE value:

Both Linear Regression and Multivariate Polynomial Regression have very similar MAPE of 34, suggesting that these models have similar performance in terms of percentage error.
As, lower value indicates better performance, Multivariate Polynomial Regression performs the best.

</h1>

<br>


<h2> 
3. Calculating MAE for each model. <br>
#Mean Absolute Error
</h2>


In [107]:
# Calculate MAE for each model
mae_lin_reg = mean_absolute_error(y_test, y_pred_lin_reg)
mae_bmi_poly = mean_absolute_error(y_val, y_pred_bmi_poly)
mae_multi_poly = mean_absolute_error(y_val, y_pred_multi_poly)

print("\nMAE (Linear Regression):", mae_lin_reg)
print("MAE (BMI Polynomial Regression):", mae_bmi_poly)
print("MAE (Multivariate Polynomial Regression):", mae_multi_poly)



MAE (Linear Regression): 46.72401655139614
MAE (BMI Polynomial Regression): 53.32295802979787
MAE (Multivariate Polynomial Regression): 38.22127405114377


<h1> Based on MAE value: <br>
In terms of MAE, lower values are better, as they indicate that the model's predictions are closer to the actual values. so,

The Multivariate Polynomial Regression model has the lowest MAE (38.22), which means it has the best performance in terms of absolute prediction errors on the validation data.



</h1>

<br>


<h4>

Please answer the following questions:
How many parameters are we fitting for each of the three models? Explain these values. 

</h4>

In [108]:
num_params_linear = x_train.shape[1] + 1  # Features + 1 for the intercept in Linear Regression
num_params_bmi_poly = x_train_bmi_poly.shape[1]  # Parameters for BMI Polynomial Regression
num_params_multi_poly = X_train_poly.shape[1] 


print(f"Number of parameters (Linear Regression): {num_params_linear}")
print(f"Number of parameters (BMI Polynomial Regression): {num_params_bmi_poly}")
print(f"Number of parameters (Multivariate Polynomial Regression): {num_params_multi_poly}")

Number of parameters (Linear Regression): 11
Number of parameters (BMI Polynomial Regression): 3
Number of parameters (Multivariate Polynomial Regression): 65


<h3> <i>Which model would you choose for deployment, and why? (1 point) </i> </h3>
<br>

<h1>
Based on camparing the model predicted parameters, Multivariate Polynomial Regression is the best as it has better predicition accuracy.


</h1>